File: Calto-GG/Goldman Sachs/Day 2/Number_Of_Boomerangs.py
#  For each point, we create a dictionary (cmap) to store the number of points 
# that are a certain distance away from the current point. The distance is 
# calculated using the difference in x and y coordinates, and stored as the key 
# in the dictionary.
# Then we iterate over the dictionary, and for each key, we calculate the 
# number of boomerangs that can be formed using the value of the key, 
# which is the number of points that are that distance away.
# Finally we return the total number of boomerangs.

class solution:
    def numberOfBoomerangs(self, points: list[list[int]]) -> int:
        res = 0
        # Iterate through each point in the input list
        for p in points:
            cmap = {}
            # calculate distance from current point 'p' to all other points
            for q in points:
                # calculate difference in x and y coordinates
                f = p[0]-q[0]
                s = p[1]-q[1]
                # store distance as key in dictionary, and increase count by 1
                cmap[f*f + s*s] = 1 + cmap.get(f*f + s*s, 0)
            # Iterate through the dictionary
            for k in cmap:
                # calculate number of boomerangs that can be formed using the value of the key
                res += cmap[k] * (cmap[k] -1)
        return res

File: Shubham Garad/Goldman Sachs/Day7/NumberOfBoomerangs.cpp
class Solution {
public:
    int numberOfBoomerangs(vector<vector<int>>& points) {
        
        int count =0;
        int n = points.size();
        if(n<3)
        {
            return 0;
        }

        for(int i=0;i<n;++i)
        {
            map<int,int>mp;
            for(int j =0 ;j<n;++j)
            {
                int x = pow(points[i][0]-points[j][0],2)+pow(points[i][1]-points[j][1],2);

                mp[x]++;
            }   
            for(auto it : mp)
            {
                if(it.second)
                {
                    count += it.second*(it.second-1);
                }
            }
        }

        return count;
    }
};